# Secrets of The Dark Forest

- Автор: Бекишев С.В.
- Дата: 12.11.2025

### Цели и задачи проекта

<font color='#777778'>Перечислите цель проекта и задачи, которые вы решаете. Для этого можно использовать описание проекта, но полезнее будет сформулировать основную цель проекта самостоятельно.</font>

### Описание данных

<font color='#777778'>Здесь приведите описание данных.</font>

### Содержимое проекта

<font color='#777778'>Перечислите основные шаги проекта или напишите оглавление. Используйте описание проекта, чтобы зафиксировать основные шаги.</font>

---

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [7]:
import pandas as pd

In [8]:
df = pd.read_csv('new_games.csv')
df

Name Platform  Year of Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       NA sales EU sales JP sales  Other sales  Critic Score User Score Rating  
0         41.36    28.96     3.77         8.45          76.0          8      E  
1         29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2         15.68    12.76     3.79         3.29          82.0        8.3      E  
3         15.61    10.93     3.28         2.95          80.0          8      E  
4         11.27     8.89    10.22         1.00           NaN        NaN    NaN  
...         ...      ...      ...          ...           ...        ...    ...  
16951      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16952      0.00     0.01      0.0         0.00           NaN        NaN    NaN  
16953      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16954      0.01      0.0      0.0         0.00           NaN        NaN    NaN  
16955      0.00      0.0     0.01         0.00           NaN        NaN    NaN  

[16956 rows x 11 columns]

In [9]:
df.describe()

Year of Release      NA sales   Other sales  Critic Score
count     16681.000000  16956.000000  16956.000000   8242.000000
mean       2006.485522      0.262023      0.047087     68.926717
std           5.873102      0.808654      0.185577     13.944565
min        1980.000000      0.000000      0.000000     13.000000
25%        2003.000000      0.000000      0.000000     60.000000
50%        2007.000000      0.080000      0.010000     71.000000
75%        2010.000000      0.240000      0.030000     79.000000
max        2016.000000     41.360000     10.570000     98.000000

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


Комментарии по датафрейму:
- Существуют пропуски в столбцах:
  - Name
  - Year of Release
  - Genre
  - Critic Score: около половины значения NaN
  - User Score
  - Rating
- Типы данных:
  - Year of Release: сейчас тип float, но должен быть int
  - EU sales: сейчас тип object, но должен быть float
  - JP sales: сейчас тип object, но должен быть float
  - Critic Score: сейчас тип float, но можно сделать int
  - User Score: сейчас тип object, но можно убрать значения 'tbd', и преобразовать в float
- Названия столбцов не приведены к snake_case

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [13]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [14]:
# Преобразование названиq столбцов в snake_case
for column in df.columns:
    snake_column = column.lower().replace(' ', '_')
    df = df.rename(columns={column : snake_column})

In [15]:
# Проверка
df.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  eu_sales jp_sales  other_sales  critic_score user_score rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [17]:
# Создаем копию датафрейма для будущего сравнения с начальным
c_df = df.copy()

#### Проверка столбца 'name'

In [19]:
c_df[c_df['name'].isna() == True]

name platform  year_of_release genre  na_sales eu_sales jp_sales  \
661    NaN      GEN           1993.0   NaN      1.78     0.53      0.0   
14439  NaN      GEN           1993.0   NaN      0.00      0.0     0.03   

       other_sales  critic_score user_score rating  
661           0.08           NaN        NaN    NaN  
14439         0.00           NaN        NaN    NaN

In [20]:
c_df = c_df.dropna(subset=['name'])

In [21]:
# Расчет доли удаленных строк от начального датафрейма (в процентах)
(1 - len(c_df['name']) / len(df['name'])) * 100

0.011795234725175963

Вывод: 
- Проверяем строки где name = NaN: 
- Найденные строки имеют малую смысловую нагрузку и потому их можно удалить
- Удаленные строки имеют незначительный вес, всего ~0,012%

#### Проверка столбца platform

In [24]:
c_df['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

В столбце 'platform' никаких проблем не обнаружено. В столбце нет значений с NaN или другими 'заглушками'

#### Проеврка столбца year_of_release

In [27]:
# Строки с NaN
nan_year_of_release = c_df[c_df['year_of_release'].isna() == True]
nan_year_of_release

name platform  year_of_release  \
183                      Madden NFL 2004      PS2              NaN   
379                     FIFA Soccer 2004      PS2              NaN   
458           LEGO Batman: The Videogame      Wii              NaN   
477           wwe Smackdown vs. Raw 2006      PS2              NaN   
611                       Space Invaders     2600              NaN   
...                                  ...      ...              ...   
16609  PDC World Championship Darts 2008      PSP              NaN   
16641                      Freaky Flyers       GC              NaN   
16685                          Inversion       PC              NaN   
16695        Hakuouki: Shinsengumi Kitan      PS3              NaN   
16760                       Virtua Quest       GC              NaN   

              genre  na_sales eu_sales jp_sales  other_sales  critic_score  \
183          Sports      4.26     0.26     0.01         0.71          94.0   
379          Sports      0.59     2.36     0.04         0.51          84.0   
458          Action      1.80     0.97      0.0         0.29          74.0   
477        Fighting      1.57     1.02      0.0         0.41           NaN   
611         Shooter      2.36     0.14      0.0         0.03           NaN   
...             ...       ...      ...      ...          ...           ...   
16609        Sports      0.01      0.0      0.0         0.00          43.0   
16641        Racing      0.01      0.0      0.0         0.00          69.0   
16685       Shooter      0.01      0.0      0.0         0.00          59.0   
16695     Adventure      0.01      0.0      0.0         0.00           NaN   
16760  Role-Playing      0.01      0.0      0.0         0.00          55.0   

      user_score rating  
183          8.5      E  
379          6.4      E  
458          7.9   E10+  
477          NaN    NaN  
611          NaN    NaN  
...          ...    ...  
16609        tbd   E10+  
16641        6.5      T  
16685        6.7      M  
16695        NaN    NaN  
16760        5.5      T  

[275 rows x 11 columns]

In [28]:
# Расчет доли от текущего датафрема где year_of_release = NaN (в процентах)
round(len(nan_year_of_release) / len(df) * 100, 3)

1.622

In [29]:
# Количество строк где year_of_release = NaN
len(c_df[c_df['year_of_release'].isna() == True])

275

In [30]:
# Для игр где year_of_release = NaN, но в названии есть приписка о годе, подстваляем этот год в year_of_release
# Подобная вставка может не отражать точный год, но скорее всего отклонение будет небольшое: +- 1 год

def check_year(row):
    try:
        year = int(row['name'][-4:])
        # Проверяем, что это год
        if year >= 1960 and year <= 2025:
            return year
        else:
            return row['year_of_release']
    except:
        return row['year_of_release']

c_df['year_of_release'] = c_df.apply(check_year, axis=1)
c_df

name platform  year_of_release         genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2007.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       na_sales eu_sales jp_sales  other_sales  critic_score user_score rating  
0         41.36    28.96     3.77         8.45          76.0          8      E  
1         29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2         15.68    12.76     3.79         3.29          82.0        8.3      E  
3         15.61    10.93     3.28         2.95          80.0          8      E  
4         11.27     8.89    10.22         1.00           NaN        NaN    NaN  
...         ...      ...      ...          ...           ...        ...    ...  
16951      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16952      0.00     0.01      0.0         0.00           NaN        NaN    NaN  
16953      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16954      0.01      0.0      0.0         0.00           NaN        NaN    NaN  
16955      0.00      0.0     0.01         0.00           NaN        NaN    NaN  

[16954 rows x 11 columns]

In [31]:
print(f"Количество строк с NaN в столбце year_of_release: {len(c_df[c_df['year_of_release'].isna() == True])}")

Количество строк с NaN в столбце year_of_release: 260


In [32]:
# Заполнение средним значением года по платформе для всех оставшихся year_of_release = NaN
def year_by_platform(row):
    if pd.isna(row['year_of_release']):
        group = c_df[c_df['platform'] == row['platform']]
        return round(group['year_of_release'].mean(), 0)
    else:
        return row['year_of_release']

In [33]:
c_df['year_of_release'] = c_df.apply(year_by_platform, axis=1)

# Меняем тип данных с float на int
c_df['year_of_release'] = c_df['year_of_release'].astype('int64')

# Проверка
c_df['year_of_release'].dtypes

dtype('int64')

Выводы:
- В изначальном датафрейме было 275 строк с year_of_release = NaN, что равно ~1.622% от общего кол-ва строк
- В некоторых строках было замечено, что в конце некоторых строк столбца 'name' содержится указание на год, такой год может не отражать точное значение, но часто такие указния имеют небольшое отклонение от реального года релиза. Таких строк нашлось немного - всего 15.
- Для всех остальных строк year_of_release = NaN было установлено среднее значение года релиза в зависимости от платформы

#### Проверка столбца genre

In [36]:
print(f"Количество NaN значений в столбце 'genre': {len(c_df[c_df['genre'].isna() == True])}\n")
print(c_df['genre'].unique())

Количество NaN значений в столбце 'genre': 0

['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']


В столбце 'genre' никаких проблем не обнаружено.

#### Проверка столбца eu_sales

Так как в столбце 'na sales' количество ненулевых строк совпадает с количеством строк во всем датафрейме и тип данных установлен верно (float), предполагаем, что с этим столбцов все в порядке

In [40]:
# Пытаемся преобразовать к нужному, тем самым проверяя есть ли неподходящие значения
# c_df['eu_sales'] = c_df['eu_sales'].astype('float64')

In [41]:
# Выводим строки с "unknown"
print(f"Количество строк с 'unknown' = {len(c_df[c_df['eu_sales']=='unknown'])}")
c_df[c_df['eu_sales']=='unknown']

Количество строк с 'unknown' = 6


name platform  year_of_release     genre  \
446                      Rhythm Heaven       DS             2008      Misc   
802                        Dead Rising     X360             2006    Action   
1131  Prince of Persia: Warrior Within      PS2             2004    Action   
1132                         Far Cry 4     XOne             2014   Shooter   
1394                   Sonic Advance 3      GBA             2004  Platform   
1612                       Ratatouille       DS             2007    Action   

      na_sales eu_sales jp_sales  other_sales  critic_score user_score rating  
446       0.55  unknown     1.93         0.13          83.0          9      E  
802       1.16  unknown     0.08         0.20          85.0        7.6      M  
1131      0.54  unknown      0.0         0.22          83.0        8.5      M  
1132      0.80  unknown     0.01         0.14          82.0        7.5      M  
1394      0.74  unknown     0.08         0.06          79.0        8.4      E  
1612      0.49  unknown      0.0         0.14           NaN        NaN    NaN

Строк со значением 'unknown' мало, удалять их нецелесообразно, потому что в них много информации. Поэтому их можно заменить на -1

In [43]:
# Преобразуем значения 'unknown' в NaN
c_df['eu_sales'] = pd.to_numeric(c_df['eu_sales'], errors='coerce')
# Меняем NaN на -1
c_df['eu_sales'] = c_df['eu_sales'].fillna(-1) 

# Проверка
c_df['eu_sales'].dtype

dtype('float64')

#### Проверка столбца jp_sales (Все тоже самое как и с eu_sales)

In [45]:
# Пытаемся преобразовать к нужному, тем самым проверяя есть ли неподходящие значения
# c_df['jp_sales'] = c_df['jp_sales'].astype('float64')

In [46]:
print(f"Количество строк с 'unknown' = {len(c_df[c_df['jp_sales']=='unknown'])}")
c_df[c_df['jp_sales']=='unknown']

Количество строк с 'unknown' = 4


name platform  year_of_release  \
467                            Saints Row 2     X360             2008   
819                     UFC 2009 Undisputed     X360             2009   
1379                      Hello Kitty Party       DS             2007   
4732  Castlevania: The Dracula X Chronicles      PSP             2007   

         genre  na_sales  eu_sales jp_sales  other_sales  critic_score  \
467     Action      1.94      0.79  unknown         0.28          81.0   
819   Fighting      1.48      0.39  unknown         0.19          83.0   
1379      Misc      0.78      0.51  unknown         0.12           NaN   
4732  Platform      0.22      0.09  unknown         0.07          80.0   

     user_score rating  
467         8.1      M  
819         7.9      T  
1379        tbd      E  
4732        7.8      T

In [47]:
# Преобразуем значения 'unknown' в NaN
c_df['jp_sales'] = pd.to_numeric(c_df['jp_sales'], errors='coerce')
# Меняем NaN на -1
c_df['jp_sales'] = c_df['jp_sales'].fillna(-1) 

# Проверка
c_df['eu_sales'].dtype

dtype('float64')